crypto/evp/evp_rand.c

```diff
@@ -0,0 +1,532 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "internal/cryptlib.h"
+#include <openssl/engine.h>
+#include <openssl/evp.h>
+#include <openssl/x509v3.h>
+#include <openssl/rand.h>
+#include <openssl/core.h>
+#include <openssl/core_names.h>
+#include <openssl/crypto.h>
+#include "crypto/asn1.h"
+#include "crypto/evp.h"
+#include "internal/numbers.h"
+#include "internal/provider.h"
+#include "evp_local.h"
+
+static int evp_rand_up_ref(void *vrand)
+{
+    EVP_RAND *rand = (EVP_RAND *)vrand;
+    int ref = 0;
+
+    if (rand != NULL)
+        return CRYPTO_UP_REF(&rand->refcnt, &ref, rand->lock);
+    return 1;
+}
+
+static void evp_rand_free(void *vrand){
+    EVP_RAND *rand = (EVP_RAND *)vrand;
+    int ref = 0;
+
+    if (rand != NULL) {
+        CRYPTO_DOWN_REF(&rand->refcnt, &ref, rand->lock);
+        if (ref <= 0) {
+            ossl_provider_free(rand->prov);
+            CRYPTO_THREAD_lock_free(rand->lock);
+            OPENSSL_free(rand);
+        }
+    }
+}
+
+static void *evp_rand_new(void)
+{
+    EVP_RAND *rand = NULL;
+
+    if ((rand = OPENSSL_zalloc(sizeof(*rand))) == NULL
+        || (rand->lock = CRYPTO_THREAD_lock_new()) == NULL) {
+        evp_rand_free(rand);
+        return NULL;
+    }
+    rand->refcnt = 1;
+    return rand;
+}
+
+/* Enable locking of the underlying DRBG/RAND if available */
+int EVP_RAND_CTX_enable_locking(EVP_RAND_CTX *rand)
+{
+    if (rand->meth->enable_prov_locking != NULL)
+        return rand->meth->enable_prov_locking(rand->data);
+    return 1;
+}
+
+/* Lock the underlying DRBG/RAND if available */
+static int evp_rand_lock(EVP_RAND_CTX *rand)
+{
+    if (rand->meth->prov_lock != NULL)
+        return rand->meth->prov_lock(rand->data);
+    return 1;
+}
+
+/* Unlock the underlying DRBG/RAND if available */
+static void evp_rand_unlock(EVP_RAND_CTX *rand)
```

Comment:
history has shown has that void functions are often a mistake.  return a status.

Reply:
What's possible recovery is there if the unlock fails?  There are some calls where an error return just doesn't make sense: close or free e.g.

Plus, this is a local function.

Reply:
> There are some calls where an error return just doesn't make sense: close or free e.g.

Hi, you've triggered a personal bugbear of mine: the error return from close.  For network file systems you'll typically have some metadata to write at close() time (if not actual data that's been cached locally for performance).  If the fileserver has gone down in the interim, an error from close() is the only signal possible to the application that their data is at risk.  While many applications don't actually check, they do so at their own (well, their user's data's) risk.
</soapbox>

Reply:
Says the maintainer of AFS :)

Reply:
Okay, close wasn't an ideal example.  I'm not sure how an application should deal with it failing.  Try again?  Forever?  Leak the descriptor until exit?

An application that knows it is using a network file systems could perhaps do something.

Reply:
> Says the maintainer of AFS :)

That may have a lot to do with why it's a personal bugbear of mine, yes :)

Reply:
I would argue that making it look to the called like the unlock succeeded (`void` can be interpreted as "this never fails") leads the caller astray, probably to crash mysteriously somewhere else.  If nothing else, that makes a failure in this function (or in `rand->meth->prov_unlock()`) very hard to find, should one need to debug.

Reply:
Returning a status here may not allow recovery, but at least allows the caller to know that something went wrong, and to abort in some way (gracefully or not)